common/middleware/src/meta_rules.rs

```diff
@@ -0,0 +1,516 @@
+use http::types::body::{KeyValueMap, LineBufferMut};
+use std::collections::HashMap;
+
+/// Env config options
+static LOGDNA_META_APP: &str = "LOGDNA_META_APP";
+static LOGDNA_META_HOST: &str = "LOGDNA_META_HOST";
+static LOGDNA_META_ENV: &str = "LOGDNA_META_ENV";
+static LOGDNA_META_FILE: &str = "LOGDNA_META_FILE";
+static LOGDNA_META_K8S_FILE: &str = "LOGDNA_META_K8S_FILE";
+static LOGDNA_META_JSON: &str = "LOGDNA_META_JSON";
+static LOGDNA_META_ANNOTATIONS: &str = "LOGDNA_META_ANNOTATIONS";
+static LOGDNA_META_LABELS: &str = "LOGDNA_META_LABELS";
+
+//TODO: extract to LogConfig
+pub struct MetaRulesConfig {
+    pub app: Option<String>,
+    pub host: Option<String>,
+    pub env: Option<String>,
+    pub file: Option<String>,
+    pub k8s_file: Option<String>, // for k8s lines, applied after "file"
+    pub meta: Option<String>,
+    pub annotations: Option<String>,
+    pub labels: Option<String>,
+}
+
+impl MetaRulesConfig {
+    pub fn default() -> Self {
+        MetaRulesConfig {
+            app: None,
+            host: None,
+            env: None,
+            file: None,
+            k8s_file: None,
+            meta: None,
+            annotations: None,
+            labels: None,
+        }
+    }
+
+    pub fn from_env() -> Self {
+        let vars = os_env_hashmap();
+        MetaRulesConfig {
+            app: vars.get(LOGDNA_META_APP).cloned(),
+            host: vars.get(LOGDNA_META_HOST).cloned(),
+            env: vars.get(LOGDNA_META_ENV).cloned(),
+            file: vars.get(LOGDNA_META_FILE).cloned(),
+            k8s_file: vars.get(LOGDNA_META_K8S_FILE).cloned(),
+            meta: vars.get(LOGDNA_META_JSON).cloned(),
+            annotations: vars.get(LOGDNA_META_ANNOTATIONS).cloned(),
+            labels: vars.get(LOGDNA_META_LABELS).cloned(),
+        }
+    }
+}
+
+pub struct MetaRules {
+    env_map: HashMap<String, String>,
+    // "override" fields
+    over_app: Option<String>,
+    over_host: Option<String>,
+    over_env: Option<String>,
+    over_file: Option<String>,
+    over_k8s_file: Option<String>, // k8s lines only
+    over_meta: Option<String>,
+    over_annotations: Option<HashMap<String, String>>, // "merge override", for  "delete and then override" - disable k8s enrichment
+    over_labels: Option<HashMap<String, String>>,      // --//--
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum MetaRulesError {
+    #[error(transparent)]
+    Serde(#[from] serde_json::Error),
+}
+
+impl MetaRules {
+    pub fn new(cfg: MetaRulesConfig) -> Result<MetaRules, MetaRulesError> {
+        let obj = MetaRules {
+            env_map: os_env_hashmap(),
+            over_app: cfg.app,
+            over_host: cfg.host,
+            over_env: cfg.env,
+            over_file: cfg.file,
+            over_k8s_file: cfg.k8s_file,
+            over_meta: cfg.meta,
+            over_annotations: cfg
+                .annotations
+                .map_or_else(|| None, |str| serde_json::from_str(str.as_str()).unwrap()),
```

Comment:
`.map_or_else(|| None,...)` is just `map(...)`, right?

Unwrapping on user provides data seems dangerous. Could we change these lines to something like the below in order to propagate the serde error?

```
            over_annotations: cfg
                .annotations
                .map(|data| serde_json::from_str(data.as_str()))
                .transpose()?
```

or even

```
            over_annotations: cfg
                .annotations.as_deref()
                .map(serde_json::from_str)
                .transpose()?
```
If we wanted to do it points free :)

Reply:
+1. updating.